ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -501,6 +503,38 @@ static int HandleVnodeOperation(
                     goto CleanupAndReturn;
                 }
             }
+            
+            if (ActionBitIsSet(action, KAUTH_VNODE_WRITE_DATA))
```

Comment:
I'm not 100% clear on when exactly we need the `MessageType_KtoU_NotifyPreConvertToFull` notification so this could be intentional, but I notice that if the file is initially empty and we get a `KAUTH_VNODE_WRITE_DATA` authorisation callback, we will go through the logic for hydrating the file (tell the provider `MessageType_KtoU_HydrateFile`) just above this new case. Would we in this case not *also* want to send the `MessageType_KtoU_NotifyPreConvertToFull` message once the file is hydrated? Currently that won't happen due to the `goto CleanupAndReturn;` above.

Reply:
No, we don't because that will cause git to start tracking the file and in the case of only reading a file we don't want git to start tracking it as it will greatly impact the performance of git commands.
We only want git to start tracking and checking files for changes once a write has happened since that is a lot less common than reads and git will stay fast because of the limited number of files it has to go through.

Reply:
> We only want git to start tracking and checking files for changes once a write has happened

Sure, but when we get `KAUTH_VNODE_WRITE_DATA`, the write is about to happen, so why not if the file was previously empty, only if it was already hydrated at that point?

Reply:
If I understand correctly, in the case of a write event and the file is empty, the code needs to hydrate the file and mark it as full, which is not happening currently with the `goto` above after the file is hydrated.  Isn't the `goto` only in error conditions?  Sorry I'm not yet familiar with this code.

Reply:
Ah, sorry - of course, the goto only happens in the error condition. My mistake.